From 7204bd35996253c90ab0cbdd6ca4a8b0941e06cf Mon Sep 17 00:00:00 2001
From: Eugene Mutavchi <Ievgen_Mutavchi@comcast.com>
Date: Wed, 30 Jan 2019 09:26:24 -0500
Subject: [PATCH] Fix dispatching of progress events in paused state

And use video buffer size limits for audio only playbacks
---
 Source/WebCore/Modules/mediasource/SourceBuffer.cpp              | 3 +++
 .../graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp    | 9 +++++++++
 .../graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.h      | 1 +
 3 files changed, 13 insertions(+)

diff --git a/Source/WebCore/Modules/mediasource/SourceBuffer.cpp b/Source/WebCore/Modules/mediasource/SourceBuffer.cpp
index 0b208c6..a141b41 100644
--- a/Source/WebCore/Modules/mediasource/SourceBuffer.cpp
+++ b/Source/WebCore/Modules/mediasource/SourceBuffer.cpp
@@ -1140,6 +1140,9 @@ size_t SourceBuffer::maximumBufferSize() const
     if (!maxBufferSizeVideo)
         maximumBufferSizeDefaults(maxBufferSizeVideo, maxBufferSizeAudio, maxBufferSizeText);
 
+    if (m_source && m_source->sourceBuffers() && m_source->sourceBuffers()->length() == 1)
+        return maxBufferSizeVideo;
+
     if (m_videoTracks && m_videoTracks->length() > 0)
         return maxBufferSizeVideo;
     if (m_audioTracks && m_audioTracks->length() > 0)
diff --git a/Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp b/Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp
index ed609e6..e90669d 100644
--- a/Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp
+++ b/Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp
@@ -1251,6 +1251,15 @@ std::optional<PlatformVideoPlaybackQualityMetrics> MediaPlayerPrivateGStreamerMS
     return std::nullopt;
 }
 
+MediaTime MediaPlayerPrivateGStreamerMSE::maxTimeLoaded() const
+{
+    if (m_errorOccured)
+        return MediaTime::zeroTime();
+
+    MediaTime maxBufferedTime = buffered()->maximumBufferedTime();
+    return maxBufferedTime.isValid() ? maxBufferedTime : MediaTime::zeroTime();
+}
+
 } // namespace WebCore.
 
 #endif // USE(GSTREAMER)
diff --git a/Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.h b/Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.h
index 3bdc85e..c8c4346 100644
--- a/Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.h
+++ b/Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.h
@@ -93,6 +93,7 @@ public:
     void dispatchDecryptionStructure(GUniquePtr<GstStructure>&&) final;
 #endif
 
+    MediaTime maxTimeLoaded() const override;
 private:
     static void getSupportedTypes(HashSet<String, ASCIICaseInsensitiveHash>&);
     static MediaPlayer::SupportsType supportsType(const MediaEngineSupportParameters&);
-- 
2.7.4

